src/v/raft/group_manager.cc

```diff
@@ -190,30 +189,25 @@ raft::group_configuration group_manager::create_initial_configuration(
 }
 
 ss::future<> group_manager::remove(ss::lw_shared_ptr<raft::consensus> c) {
-    return _groups_mutex.with([this, c = std::move(c)] {
-        return c->stop()
-          .discard_result()
-          .then([c] { return c->remove_persistent_state(); })
-          .then([this, id = c->group()] {
-              return _heartbeats.deregister_group(id);
-          })
-          .finally([this, c] {
-              _groups.erase(
-                std::remove(_groups.begin(), _groups.end(), c), _groups.end());
-          });
-    });
+    return do_shutdown(std::move(c), true).discard_result();
 }
 
 ss::future<xshard_transfer_state>
 group_manager::shutdown(ss::lw_shared_ptr<raft::consensus> c) {
-    auto units = co_await _groups_mutex.get_units();
-
-    auto xst_state = co_await c->stop();
-    co_await _heartbeats.deregister_group(c->group());
-    _groups.erase(
-      std::remove(_groups.begin(), _groups.end(), c), _groups.end());
+    return do_shutdown(std::move(c), false);
+}
 
-    co_return xst_state;
+ss::future<xshard_transfer_state> group_manager::do_shutdown(
+  ss::lw_shared_ptr<raft::consensus> c, bool remove_persistent_state) {
+    const auto group_id = c->group();
+    auto transfer_state = co_await c->stop();
+    if (remove_persistent_state) {
+        co_await c->remove_persistent_state();
+    }
+    co_await _heartbeats.deregister_group(group_id);
+    auto it = std::find(_groups.begin(), _groups.end(), c);
+    _groups.erase(it);
```

Comment:
should we assert that `it` is not `_groups.end()`? I think it's ub to use the end iterator here.

Reply:
>should we assert that it is not _groups.end()? I think it's ub to use the end iterator here.

the question i had @mmaslankaprv is what prevents group::remove or group::shutdown from being called more than once for the same group? that seems to be what prevents this assertion from triggering, but i'm also not sure how that guarantee is made, especially since there are two different entry points that end up in do_shutdown.